### DISCUSSION TITLE
Error in startup script: invalid command name "wm"

### DISCUSSION THREAD

**Discussion by ergauravjain** [TPC-Council/HammerDB#470] [Q&A] [answered]
HI Team ,
i need your help here i am trying to setup the hammer db in teh linux server and try to test only mysql when i run ./hammerdb command its not getting trough with nay vertion its have tested with 4.4. 4.5 and 4.6

[root@svl-dbp657 HammerDB-4.6]# ./hammerdb
application-specific initialization failed: no display name and no $DISPLAY environment variable
Error in startup script: invalid command name "wm"
    while executing
"wm title            . "Loading HammerDB""
    (in namespace eval "::LoadingProgressMeter" script line 4)
    invoked from within
"namespace eval LoadingProgressMeter {
    set max 14

    wm title            . "Loading HammerDB"
    wm protocol         . WM_DELETE_WINDOW {#Do not..."
    (file "./hammerdb" line 54)
can some help here to setup my environment .

also when i am trying to setup with windows its not able to find the client library for my sql however its already part of the location /

 couldn't load library "D:/MSSQL/HammerDB/HammerDB-4.5-Win/HammerDB-4.5/lib/mysqltcl3.052/libmysqltcl.dll": this library or a dependent library could not be found in library path

**Comment by sm-shaw** [comment]
Hi, this will be down to the ownership of the DISPLAY so as an example '$' shows the user prompt and '#' root:
So firstly we are logged in as the user and check the DISPLAY and run HammerDB
```
$ echo $DISPLAY
:0
$ ./hammerdb
```
It all works and HammerDB opens. 
Now just as an example, we change the name of the DISPLAY and try the same thing:
```
$ export DISPLAY=notmydisplay:0
$ ./hammerdb
application-specific initialization failed: couldn't connect to display "notmydisplay:0"
Error in startup script: invalid command name "wm"
...
```
This errored because we can't connect to the DISPLAY. 
So now we su to root and try to run HammerDB and we see the same error as you see:
```
# ./hammerdb
application-specific initialization failed: no display name and no $DISPLAY environment variable
Error in startup script: invalid command name "wm"
 ...
```
So we set the DISPLAY environment variable, but still it doesn't work
```
# export DISPLAY=:0
root@CRANE:/home/steve/HammerDB-4.6# ./hammerdb
No protocol specified
No protocol specified
application-specific initialization failed: couldn't connect to display ":0"
Error in startup script: invalid command name "wm"
...
```
So now we go back to our user prompt (the user that logged in and owns the display and allow other clients to connect
```
$ export DISPLAY=:0
$ xhost +
access control disabled, clients can connect from any host
```
and back as root, success! we now have both the environment variable set and the permission to run the GUI as root.
```
# export DISPLAY=:0
# ./hammerdb
```
For the error on Windows
`couldn't load library "D:/MSSQL/HammerDB/HammerDB-4.5-Win/HammerDB-4.5/lib/mysqltcl3.052/libmysqltcl.dll": this library or a dependent library could not be found in library path`

See this section of the documentation https://www.hammerdb.com/docs/ch01s10.html. This explains step by step using MySQL as an example of how to verify that the MySQL library that HammerDB is looking for is installed.

**Comment by ergauravjain** [comment]
HI ,

Thanks for your response still i have same issue and i am using root only .

 HammerDB-4.6]# echo $DISPLAY

 export DISPLAY=notmydisplay:0
 ./hammerdb
application-specific initialization failed: couldn't connect to display "notmydisplay:0"
Error in startup script: invalid command name "wm"
    while executing
"wm title            . "Loading HammerDB""
    (in namespace eval "::LoadingProgressMeter" script line 4)
    invoked from within
"namespace eval LoadingProgressMeter {
    set max 14

    wm title            . "Loading HammerDB"
    wm protocol         . WM_DELETE_WINDOW {#Do not..."
    (file "./hammerdb" line 54)
# echo $DISPLAY
notmydisplay:0
[root@ HammerDB-4.6]# ./hammerdb
application-specific initialization failed: couldn't connect to display "notmydisplay:0"
Error in startup script: invalid command name "wm"
    while executing
"wm title            . "Loading HammerDB""
    (in namespace eval "::LoadingProgressMeter" script line 4)
    invoked from within
"namespace eval LoadingProgressMeter {
    set max 14

    wm title            . "Loading HammerDB"
    wm protocol         . WM_DELETE_WINDOW {#Do not..."
    (file "./hammerdb" line 54)
[root@ HammerDB-4.6]# export DISPLAY=:0
[root@HammerDB-4.6]# ./hammerdb
application-specific initialization failed: couldn't connect to display ":0"
Error in startup script: invalid command name "wm"
    while executing
"wm title            . "Loading HammerDB""
    (in namespace eval "::LoadingProgressMeter" script line 4)
    invoked from within
"namespace eval LoadingProgressMeter {
    set max 14

    wm title            . "Loading HammerDB"
    wm protocol         . WM_DELETE_WINDOW {#Do not..."
    (file "./hammerdb" line 54)
[root@HammerDB-4.6]# export DISPLAY=:0
[root@ HammerDB-4.6]# xhost +
xhost:  unable to open display ":0"
[root@HammerDB-4.6]# ./hammerdb
application-specific initialization failed: couldn't connect to display ":0"
Error in startup script: invalid command name "wm"
    while executing
"wm title            . "Loading HammerDB""
    (in namespace eval "::LoadingProgressMeter" script line 4)
    invoked from within
"namespace eval LoadingProgressMeter {
    set max 14

    wm title            . "Loading HammerDB"
    wm protocol         . WM_DELETE_WINDOW {#Do not..."
    (file "./hammerdb" line 54)

**Comment by ergauravjain** [comment, marked as answer]
HI ,
its also required to load all library t work with i need work only for MySQL at this movement.
hammerdb>librarycheck
Checking database library for Oracle
Error: failed to load Oratcl - can't read "env(ORACLE_HOME)": no such variable
Ensure that Oracle client libraries are installed and the location in the LD_LIBRARY_PATH environment variable
Checking database library for MSSQLServer
Success ... loaded library tdbc::odbc for MSSQLServer
Checking database library for Db2
Error: failed to load db2tcl - couldn't load file "/var/lib/mysql/HammerDB-4.6/lib/db2tcl2.0.1/libdb2tcl.so": libdb2.so.1: cannot open shared object file: No such file or directory
Ensure that Db2 client libraries are installed and the location in the LD_LIBRARY_PATH environment variable
Checking database library for MySQL
Success ... loaded library mysqltcl for MySQL
Checking database library for PostgreSQL
Error: failed to load Pgtcl - couldn't load file "/var/lib/mysql/HammerDB-4.6/lib/pgtcl2.1.1/libpgtcl2.1.1.so": libpq.so.5: cannot open shared object file: No such file or directory
Ensure that PostgreSQL client libraries are installed and the location in the LD_LIBRARY_PATH environment variable
Checking database library for MariaDB
Success ... loaded library mariatcl for MariaDB

also i have still not able to figure out why couldn't connect to display "notmydisplay:0" over the linux .

alos i have check with dependency toll on the window its not able to find teh libmysql.ddl however complete package of mysql already installed on the server and libmysqltcl.dll already exist on teh server.

also when we are setting with the MySQL connection its required to keep the ssl over the local client because its asking the local path on the MySQL client.

**Reply by sm-shaw** [reply]
OK here is a graphical example.  The user "hammerdb" owns the display root doesn't. Whoever is logged in owns the display. In root we try and run HammerDB but DISPLAY is not set and we don't have permission.  So we give permission from the owner, set DISPLAY in the root session and then it works. (notmydisplay:0 is an example where it doesn't work)

![display](https://user-images.githubusercontent.com/38044085/206678274-db40a379-e871-4ccf-a983-4d9734bfbcef.PNG)

For the MySQL library it is full detailed how to do this in the documentation section linked above. Install the library set the PATH and then check it with a dependencies tool if you need to.

**Comment by ergauravjain** [comment]
HI ,
Thanks for promptreposne , i have tried to export with login user , root and other user with display available still its not working for me plz see the sanp.

![image](https://user-images.githubusercontent.com/72666693/206857556-a9402ccd-ea0a-4856-a12b-7f7884f2d7e3.png)

**Reply by sm-shaw** [reply]
Hi, So are you actually running a GUI? i.e. A Display server with a Desktop Environment - you will need a desktop running as per my example to be able to display a GUI. See the separate terminal windows. This can either be local or remote, a remote one can even run on a Windows environment e.g. http://www.straightrunning.com/XmingNotes/.  Note if running remote consideration needs to be given to latency and running over a long distance needs a solution such as VNC. 

 If you do not have a GUI you need to run the CLI with the command hammerdbcli to then run HammerDB workloads with text interaction.
